api/sendfild/service/service.go

```diff
@@ -0,0 +1,713 @@
+		Total int64   `bson:"total"`
+		Avg   float64 `bson:"avg"`
+		Min   int64   `bson:"min"`
+		Max   int64   `bson:"max"`
+	}
+	type report struct {
+		All            []entityCount `bson:"all"`
+		ByMessageState []entityCount `bson:"by_message_state"`
+		Waiting        []entityCount `bson:"waiting"`
+		UniqueFrom     []entityCount `bson:"unique_from"`
+		UniqueTo       []entityCount `bson:"unique_to"`
+		SentFilStats   []stats       `bson:"sent_fil_stats"`
+	}
+
+	createdAtMatch := bson.M{}
+	if req.Before != nil {
+		createdAtMatch["$lt"] = req.Before.AsTime()
+	}
+	if req.After != nil {
+		createdAtMatch["$gt"] = req.After.AsTime()
+	}
+	match := bson.M{}
+	if len(createdAtMatch) > 0 {
+		match["created_at"] = createdAtMatch
+	}
+	cursor, err := s.col.Aggregate(ctx, bson.A{
+		bson.M{"$match": match},
+		bson.M{"$facet": bson.M{
+			"all": bson.A{
+				bson.M{"$group": bson.M{"_id": nil, "count": bson.M{"$sum": 1}}},
+			},
+			"by_message_state": bson.A{
+				bson.M{"$group": bson.M{"_id": "$message_state", "count": bson.M{"$sum": 1}}},
+			},
+			"waiting": bson.A{
+				bson.M{"$match": bson.M{"waiting": true}},
+				bson.M{"$group": bson.M{"_id": nil, "count": bson.M{"$sum": 1}}},
+			},
+			"unique_from": bson.A{
+				bson.M{"$group": bson.M{"_id": "$from", "count": bson.M{"$sum": 1}}},
+			},
+			"unique_to": bson.A{
+				bson.M{"$group": bson.M{"_id": "$to", "count": bson.M{"$sum": 1}}},
+			},
+			"sent_fil_stats": bson.A{
+				bson.M{"$group": bson.M{
+					"_id":   nil,
+					"total": bson.M{"$sum": "$amount_nano_fil"},
+					"avg":   bson.M{"$avg": "$amount_nano_fil"},
+					"max":   bson.M{"$max": "$amount_nano_fil"},
+					"min":   bson.M{"$min": "$amount_nano_fil"},
+				}},
+			},
+		}},
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "calling aggregate: %v", err)
+	}
+	var res []report
+	if err = cursor.All(ctx, &res); err != nil {
+		return nil, status.Errorf(codes.Internal, "decoding cursor results: %v", err)
+	}
+	if len(res) != 1 {
+		return nil, status.Errorf(codes.Internal, "unexpected number of aggregate results: %v", len(res))
+	}
+
+	r := res[0]
+
+	resp := &pb.SummaryResponse{}
+
+	if len(r.All) == 1 {
+		resp.CountTxns = r.All[0].Count
+	}
+
+	for _, state := range r.ByMessageState {
+		switch pb.MessageState(state.ID.(int32)) {
+		case pb.MessageState_MESSAGE_STATE_PENDING:
+			resp.CountPending = state.Count
+		case pb.MessageState_MESSAGE_STATE_ACTIVE:
+			resp.CountActive = state.Count
+		case pb.MessageState_MESSAGE_STATE_FAILED:
+			resp.CountFailed = state.Count
+		}
+	}
+
+	if len(r.Waiting) == 1 {
+		resp.CountWaiting = r.Waiting[0].Count
+	}
+
+	resp.CountFromAddrs = int64(len(r.UniqueFrom))
+	resp.CountToAddrs = int64(len(r.UniqueTo))
+
+	if len(r.SentFilStats) == 1 {
+		resp.TotalNanoFilSent = r.SentFilStats[0].Total
+		resp.AvgNanoFilSent = r.SentFilStats[0].Avg
+		resp.MaxNanoFilSent = r.SentFilStats[0].Max
+		resp.MinNanoFilSent = r.SentFilStats[0].Min
+	}
+
+	return resp, nil
```

Comment:
We populate our response with the data that came from the aggregation and return it.